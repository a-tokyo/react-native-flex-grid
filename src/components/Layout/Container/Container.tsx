import React from 'react';
import { StyleSheet, View, useWindowDimensions } from 'react-native';

import { getConfig } from '../../../utils/grid';
import { getGridBreakpoint } from '../../../utils/responsive';

export declare interface ContainerProps
  extends React.ComponentProps<typeof View> {
  /** Fluid Container */
  fluid?: boolean;
  /** No Padding */
  noPadding?: boolean;
  /** Element to render - defaults to View */
  Element?: React.ElementType;
  /** style */
  style?: React.ComponentProps<typeof View>['style'];
}

const styles = StyleSheet.create({
  container: {
    width: '100%',
    justifySelf: 'center',
  },
  fluid: {
    maxWidth: '100%',
  },
});

/** Container [Bootstrap Docs](https://getbootstrap.com/docs/5.0/layout/containers) */
const Container = ({
  style,
  fluid,
  Element = View,
  noPadding,
  ...rest
}: ContainerProps) => {
  /** Grid config */
  const gridConfig = getConfig();
  /** screen width */
  const SCREEN_WIDTH = useWindowDimensions().width;
  /** current grid breakpoint */
  const gridBreakpoint = getGridBreakpoint(SCREEN_WIDTH);
  /** container maxWidth */
  const maxWidth = gridConfig.containerMaxWidths[gridBreakpoint];

  return (
    <Element
      style={[
        styles.container,
        noPadding
          ? null
          : {
              paddingHorizontal:
                gridConfig.containerPaddingsHorizontal[gridBreakpoint],
            },
        fluid
          ? styles.fluid
          : {
              maxWidth,
              marginHorizontal:
                typeof maxWidth === 'number'
                  ? (SCREEN_WIDTH - maxWidth) / 2
                  : undefined,
            },
        style,
      ]}
      {...rest}
    />
  );
};

export default Container;
