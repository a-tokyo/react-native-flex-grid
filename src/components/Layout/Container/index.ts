export * from './Container';

export { default } from './Container';
