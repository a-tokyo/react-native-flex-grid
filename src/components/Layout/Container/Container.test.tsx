import React from 'react';
import renderer from 'react-test-renderer';
import { ScrollView, Text } from 'react-native';

import Container from './Container';

const setupContainer = (overrideProps = {}) => <Container {...overrideProps} />;

describe('<Container />', () => {
  it('renders basic props correctly', () => {
    const wrapper = renderer.create(setupContainer());

    expect(wrapper).toMatchSnapshot();
  });

  it('renders basic props correctly - fluid', () => {
    const wrapper = renderer.create(
      setupContainer({
        children: <Text>test</Text>,
        fluid: true,
      }),
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders basic props correctly - fluid - noPadding', () => {
    const wrapper = renderer.create(
      setupContainer({
        children: <Text>test</Text>,
        fluid: true,
        noPadding: true,
      }),
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders basic props correctly - Element - style', () => {
    const wrapper = renderer.create(
      setupContainer({
        children: <Text>test</Text>,
        Element: ScrollView,
        style: {
          padding: 0,
        },
      }),
    );

    expect(wrapper).toMatchSnapshot();
  });
});
