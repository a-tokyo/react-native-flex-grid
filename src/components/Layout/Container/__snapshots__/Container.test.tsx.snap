// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container /> renders basic props correctly - Element - style 1`] = `
<RCTScrollView
  style={
    [
      {
        "justifySelf": "center",
        "width": "100%",
      },
      {
        "paddingHorizontal": 20,
      },
      {
        "marginHorizontal": undefined,
        "maxWidth": "100%",
      },
      {
        "padding": 0,
      },
    ]
  }
>
  <View>
    <Text>
      test
    </Text>
  </View>
</RCTScrollView>
`;

exports[`<Container /> renders basic props correctly - fluid - noPadding 1`] = `
<View
  style={
    [
      {
        "justifySelf": "center",
        "width": "100%",
      },
      null,
      {
        "maxWidth": "100%",
      },
      undefined,
    ]
  }
>
  <Text>
    test
  </Text>
</View>
`;

exports[`<Container /> renders basic props correctly - fluid 1`] = `
<View
  style={
    [
      {
        "justifySelf": "center",
        "width": "100%",
      },
      {
        "paddingHorizontal": 20,
      },
      {
        "maxWidth": "100%",
      },
      undefined,
    ]
  }
>
  <Text>
    test
  </Text>
</View>
`;

exports[`<Container /> renders basic props correctly 1`] = `
<View
  style={
    [
      {
        "justifySelf": "center",
        "width": "100%",
      },
      {
        "paddingHorizontal": 20,
      },
      {
        "marginHorizontal": undefined,
        "maxWidth": "100%",
      },
      undefined,
    ]
  }
/>
`;
